cpp/src/filter/filter.cu

```diff
@@ -0,0 +1,365 @@
+
+  launch_filter_kernel(kernel, filter_indices_span, columns, user_data, stream, mr);
+
+  auto num_selected = thrust::transform_reduce(rmm::exec_policy(stream),
+                                               filter_indices.begin(),
+                                               filter_indices.end(),
+                                               filter_histogram{},
+                                               cudf::size_type{0},
+                                               std::plus<cudf::size_type>{});
+
+  std::vector<std::unique_ptr<column>> filtered;
+
+  std::for_each(thrust::counting_iterator<size_t>(0),
+                thrust::counting_iterator<size_t>(columns.size()),
+                [&](auto i) {
+                  auto column      = columns[i];
+                  auto should_copy = !copy_mask.has_value() || (*copy_mask)[i];
+
+                  if (should_copy) {
+                    if (cudf::jit::is_scalar(base_column.size(), column.size())) {
+                      // broadcast scalar columns
+                      auto tiled = cudf::tile(cudf::table_view{{column}}, num_selected, stream, mr);
+                      auto tiled_columns = tiled->release();
+                      filtered.push_back(std::move(tiled_columns[0]));
+                    } else {
+                      auto d_column        = cudf::column_device_view::create(column, stream);
+                      auto filtered_column = filter_column(
+                        *d_column, num_selected, filter_indices_span.as_const(), stream, mr);
+                      filtered.push_back(std::move(filtered_column));
+                    }
+                  }
+                });
+
+  return filtered;
+}
+
+}  // namespace
+
+namespace detail {
+std::vector<std::unique_ptr<column>> filter(std::vector<column_view> const& columns,
+                                            std::string const& predicate_udf,
+                                            bool is_ptx,
+                                            std::optional<void*> user_data,
+                                            std::optional<std::vector<bool>> copy_mask,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(!columns.empty(), "Filters must have at least 1 column", std::invalid_argument);
+
+  auto const base_column = std::max_element(
+    columns.begin(), columns.end(), [](auto& a, auto& b) { return a.size() < b.size(); });
+
+  perform_checks(*base_column, columns, copy_mask);
+
+  auto filtered = filter_operation(
+    *base_column, columns, predicate_udf, is_ptx, user_data, copy_mask, stream, mr);
+
+  return filtered;
+}
+
+}  // namespace detail
+
+std::vector<std::unique_ptr<column>> filter(std::vector<column_view> const& columns,
+                                            std::string const& predicate_udf,
+                                            bool is_ptx,
```

Comment:
If it's public API, maybe worth considering a strong type like `is_ptx::YES/NO` instead of using raw boolean

Reply:
Here: https://github.com/rapidsai/cudf/pull/19300
I'll update it for filters as well once this PR is merged